internal/executor/service/job_state_reporter.go

```diff
@@ -0,0 +1,233 @@
+package service
+
+import (
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	clusterContext "github.com/armadaproject/armada/internal/executor/context"
+	domain2 "github.com/armadaproject/armada/internal/executor/domain"
+	"github.com/armadaproject/armada/internal/executor/reporter"
+	"github.com/armadaproject/armada/internal/executor/util"
+)
+
+type JobStateReporter struct {
+	eventReporter   reporter.EventReporter
+	clusterContext  clusterContext.ClusterContext
+	podIssueHandler IssueHandler
+}
+
+func NewJobStateReporter(
+	clusterContext clusterContext.ClusterContext,
+	eventReporter reporter.EventReporter,
+	podIssueHandler IssueHandler,
+) (*JobStateReporter, error) {
+	stateReporter := &JobStateReporter{
+		eventReporter:   eventReporter,
+		clusterContext:  clusterContext,
+		podIssueHandler: podIssueHandler,
+	}
+
+	_, err := clusterContext.AddPodEventHandler(stateReporter.podEventHandler())
+	if err != nil {
+		return nil, err
+	}
+
+	return stateReporter, nil
+}
+
+func (stateReporter *JobStateReporter) podEventHandler() cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod, ok := obj.(*v1.Pod)
+			if !ok {
+				log.Errorf("Failed to process pod event due to it being an unexpected type. Failed to process %+v", obj)
+				return
+			}
+			go stateReporter.reportCurrentStatus(pod)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldPod, ok := oldObj.(*v1.Pod)
+			if !ok {
+				log.Errorf("Failed to process pod event due to it being an unexpected type. Failed to process %+v", oldObj)
+				return
+			}
+			newPod, ok := newObj.(*v1.Pod)
+			if !ok {
+				log.Errorf("Failed to process pod event due to it being an unexpected type. Failed to process %+v", newObj)
+				return
+			}
+			go stateReporter.reportStatusUpdate(oldPod, newPod)
+		},
+	}
+}
+
+func (stateReporter *JobStateReporter) reportStatusUpdate(old *v1.Pod, new *v1.Pod) {
+	// Don't report status if the pod phase didn't change
+	if old.Status.Phase == new.Status.Phase {
+		return
+	}
+	// Don't report status change for pods Armada is deleting
+	// This prevents reporting JobFailed when we delete a pod - for example due to cancellation
+	if util.IsMarkedForDeletion(new) {
+		log.Infof("not sending event to report pod %s moving into phase %s as pod is marked for deletion", new.Name, new.Status.Phase)
+		return
+	}
+	stateReporter.reportCurrentStatus(new)
+}
+
+func (stateReporter *JobStateReporter) reportCurrentStatus(pod *v1.Pod) {
+	if !util.IsManagedPod(pod) {
+		return
+	}
+	if util.HasCurrentStateBeenReported(pod) {
+		return
+	}
+
+	event, err := reporter.CreateEventForCurrentState(pod, stateReporter.clusterContext.GetClusterId())
+	if err != nil {
+		log.Errorf("Failed to report event: %v", err)
+		return
+	}
+
+	if pod.Status.Phase == v1.PodFailed {
```

Comment:
The code in this if statement is largely the only new code, the rest is just split out from job_event_reporter